deepspeed/runtime/zero/stage2.py

```diff
@@ -837,8 +837,8 @@ def gradient_reduction_w_predivide(self, tensor):
 
         tensor_to_allreduce = tensor
 
-        if self.allreduce_always_fp32:
-            tensor_to_allreduce = tensor.float()
+        if self.communication_data_type is not None:
```

Comment:
== `torch.float32`

Reply:
Got it. So this is the case where we need the override. I would prefer to avoid having self.communication_data_type being None since that really means the model data type. How about the following?
```python
      if self.communication_data_type != tensor.dtype:
          tensor_to_allreduce = tensor.to(self.communication_data_type)
```
Also if self.communication_data_type is always a valid torch data type, then we could even eliminate the if condition and it will be a noop if `self.communication_data_type == tensor.dtype`.